src/Claim.ts

```diff
@@ -1,14 +1,59 @@
 /* tslint:disable:no-relative-imports */
-import * as bitcore from 'bitcore-lib'
 import * as crypto from 'crypto'
-import { canonize } from 'jsonld'
+import * as cuid from 'cuid'
+import * as jsonld from 'jsonld'
+import * as jsig from 'jsonld-signatures'
 
 import { IllegalArgumentException } from './Exceptions'
 import { Claim, ClaimAttributes, ClaimType, ClaimContext, isClaim } from './Interfaces'
 
-export const canonizeClaim = async (claim: Claim): Promise<string> => {
-  const contextualClaim = { ...ClaimContext, ...claim }
-  return canonize(contextualClaim)
+jsig.use('jsonld', jsonld)
+
+export const canonizeClaim = async (document: Claim): Promise<string> => {
+  const contextualClaim = {
+    '@context': document['@context'],
+    type: document.type,
+    issuer: document.issuer,
+    issuanceDate: document.issuanceDate,
+    claim: document.claim,
+    ...ClaimContext,
+  }
+  return jsonld.canonize(contextualClaim)
+}
+
+/*
+  signClaim = (signingOptions: any) => async (document: Claim): Promise<Claim>
+    signingOptions:
+      algorithm: 'Ed25519VerificationKey2018'
+      creator: 'data:,<base58PublicKeyValue>'
+      privateKeyBase58: base58 rendition of an Ed22159 private key
+    document: any valid Claim
+  NOTE: If you choose to use a different signing algorhithm, the other signingOption keys may differ.
+  Refer to the jsonld-signatures library for more detail
+*/
+export const signClaim = (signingOptions: any) => async (document: Claim): Promise<Claim> => {
+  if (!document.id) throw new IllegalArgumentException('Cannot sign a claim that has an empty .id field')
+  const generatedId = await getClaimId(document)
+  if (document.id !== generatedId)
+    throw new IllegalArgumentException('Cannot sign a claim whose id hasa been altered or generated incorrectly.')
+  if (signingOptions.creater === null || signingOptions.creator === '')
+    throw new IllegalArgumentException('Cannot sign a claim with an invalid creator in the signing options.')
+  return await jsig.sign(
+    {
+      '@context': document['@context'],
+      type: document.type,
+      issuer: document.issuer,
+      issuanceDate: document.issuanceDate,
+      claim: document.claim,
+      ...ClaimContext,
+    },
+    signingOptions
+  )
+}
+
+export const isValidSignature = (verifiedOptions: any = {}) => async (claim: Claim): Promise<boolean> => {
+  const results = await jsig.verify(claim, { checkNonce: cuid.isCuid, ...verifiedOptions })
```

Comment:
Correct. Unless we want to store a set of latest nonces, I'm not even sure how we would verify replay attacks.

Reply:
Is there a reason to pass this `checkNonce` function, then?

Reply:
If we don't provide a checkNonce function, verification will fail. Since poet-js is using cuid as a nonce, it's as good a test as any. Alternatives include return `true` always or check for existence.